addons/l10n_es_edi_facturae/models/account_edi_format.py

```diff
@@ -0,0 +1,465 @@
+        :param func fmt_curr:         The formatting function to use for the amount in the invoice currency
+        :return tuple:                  A tuple containing the Face items, the taxes and the invoice totals data.
+        """
+        items = []
+        totals = {
+            'total_gross_amount': 0.,
+            'total_general_discounts': 0.,
+            'total_general_surcharges': 0.,
+            'total_taxes_withheld': 0.,
+            'total_payments_on_account': 0.,
+            'amounts_withheld': 0.,
+        }
+        taxes = []
+        inv_curr = invoice.currency_id
+        eur_curr = invoice.company_currency_id
+        for line in invoice.invoice_line_ids:
+            invoice_line_values = EDI_XML_DICT_TEMPLATES['invoice_line'].copy()
+
+            total_price = line.quantity * line.price_unit * conv_rate  # This isn't a mistake
+            gross_amount = line.price_subtotal / conv_rate
+            discount = max(0., line.discount * total_price)
+            surcharge = max(0., - line.discount * total_price)  # + line.price_total - line.price_subtotal
+            totals['total_gross_amount'] += gross_amount
+            totals['total_general_discounts'] += discount
+            totals['total_general_surcharges'] += surcharge
+            base_line = {
+                'account': line.account_id,
+                'currency': line.currency_id,
+                'discount': line.discount,
+                'partner': line.partner_id,
+                'price_unit': line.price_unit,
+                'product': line.product_id,
+                'quantity': line.quantity,
+                'is_refund': line.is_refund,
+                'rate': line.currency_rate,
+                'record': line,
+                'taxes': line.tax_ids,
+                'extra_context': {},
+            }
+            taxes_output = [{
+                "TaxTypeCode": tax.l10n_es_edi_facturae_tax_type,
+                "TaxRate": f'{tax.amount:.3f}',
+                "TaxableBase": {
+                    'TotalAmount': fmt_curr(gross_amount, inv_curr.decimal_places),
+                    'EquivalentInEuros': fmt_curr(line.price_subtotal, eur_curr.decimal_places) if needs_equivalency else False,
+                },
+                "TaxAmount": {
+                    "TotalAmount": fmt_curr(tax._compute_taxes([base_line])['totals'][line.currency_id]['amount_tax'] * conv_rate, inv_curr.decimal_places),
+                    "EquivalentInEuros": fmt_curr(tax._compute_taxes([base_line])['totals'][line.currency_id]['amount_tax'], eur_curr.decimal_places),
+                },
+            } for tax in line.tax_ids]
+
+            invoice_line_values.update({
+                'ItemDescription': line.name,
+                'Quantity': line.quantity,
+                'UnitOfMeasure': line.product_uom_id.l10n_es_edi_facturae_uom_code,
+                'UnitPriceWithoutTax': fmt_curr(line.price_unit, inv_curr.decimal_places),
```

Comment:
Are you sure you can not have a higher precision for unit prices and quantities?  (suppose someone sells with higher or lower quantities)